src/Stryker.DataCollector/Stryker.DataCollector/CommunicationServer.cs

```diff
@@ -0,0 +1,130 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Threading;
+
+namespace Stryker.DataCollector
+{
+    public delegate void ConnectionEvent(object s, ConnectionEventArgs e);
+
+    public sealed class CommunicationServer : IDisposable
+    {
+        private readonly object _lck = new object();
+        private volatile bool _mustShutdown;
+        private readonly IList<CommunicationChannel> _channels = new List<CommunicationChannel>();
+        private NamedPipeServerStream _listener;
+        public string PipeName { get; }
+
+        public event ConnectionEvent RaiseNewClientEvent;
+        public event MessageReceived RaiseReceivedMessage;
+
+        private Action<string> _logger;
+
+        private static int _instanceCounter = 0;
+
+        public CommunicationServer(string name)
+        {
+            PipeName = $"StrykerPipe.{name}.{Process.GetCurrentProcess().Id}.{AppDomain.CurrentDomain.Id}:{Interlocked.Increment(ref _instanceCounter)}";
+        }
+
+        public void SetLogger(Action<string> logger)
+        {
+            _logger = logger;
+        }
+
+        public void Listen()
+        {
+            lock (_lck)
+            {
+                Log("Listen for client.");
+                if (_mustShutdown)
+                {
+                    return;
+                }
+
+                _listener = new NamedPipeServerStream(PipeName,
+                    PipeDirection.InOut,
+                    NamedPipeServerStream.MaxAllowedServerInstances,
+                    PipeTransmissionMode.Byte, PipeOptions.Asynchronous);
+                _listener.BeginWaitForConnection(OnConnect, null);
+            }
+        }
+
+        private void OnConnect(IAsyncResult ar)
+        {
+            CommunicationChannel session = null;
+            lock (_lck)
+            {
+                try
+                {
+                    _listener.EndWaitForConnection(ar);
+                }
+                catch (IOException)
+                {
+                    return;
+                }
+
+                if (_mustShutdown)
+                    return;
+
+                if (_listener.IsConnected)
+                {
+                    session = new CommunicationChannel(_listener, $"{PipeName}:S({_channels.Count})");
+                    session.SetLogger(_logger);
+                    Log($"New connection.");
+                    _channels.Add(session);
+                    _listener = null;
+                    session.RaiseReceivedMessage += Session_RaiseReceivedMessage;
+                }
+                Listen();
+            }
+            RaiseNewClientEvent?.Invoke(this, new ConnectionEventArgs(session));
+            session.Start();
+        }
+
+        private void Log(string message)
```

Comment:
log lines should be only present at verbose level. I think it can be useful to keep them, but I understand this is noise in general, hence I did put them at the lowest possible priority.

We may be able to remove them, but I would be more confortable having some production feedback before disabling.